### DISCUSSION TITLE
Pulseaudio on container

### DISCUSSION THREAD

**Discussion by jhay06** [lima-vm/lima#1107] [Q&A]
Hello would like to ask , how can i connect lima containers to default mac os audio so i can use pulseaudio on the container.
i am developing ubuntu touch apps using lima can't run desktop mode because it looks for pulse audio . then when i install pulseaudio it still error i think because pulseaudio is not routed to sound driver from mac .

**Comment by afbjorklund** [comment]
I don't think there is any sound support in the default setup, so it would have to be added as an option first.

```console
$ qemu-system-x86_64 -soundhw help
Valid sound card names (comma separated):
sb16        Creative Sound Blaster 16
es1370      ENSONIQ AudioPCI ES1370
ac97        Intel 82801AA AC97 Audio
adlib       Yamaha YM3812 (OPL2)
gus         Gravis Ultrasound GF1
cs4231a     CS4231A
hda         Intel HD Audio
pcspk       PC speaker

-soundhw all will enable all of the above
$ qemu-system-aarch64 -soundhw help
Valid sound card names (comma separated):
es1370      ENSONIQ AudioPCI ES1370
ac97        Intel 82801AA AC97 Audio
hda         Intel HD Audio

-soundhw all will enable all of the above
```

There is a newer `-audiodev` flag. Then setting up pulseaudio is a headache all by itself, once the device is there.

I believe you want the "coreaudio" output device, for Mac ?

`-device intel-hda -device hda-output,audiodev=coreaudio`

Then once you have audio working in the VM, you need to map all the required volumes (for PA) into the container.

https://github.com/mviereck/x11docker/wiki/Container-sound:-ALSA-or-Pulseaudio

**Reply by jhay06** [reply]
> I don't think there is any sound support in the default setup, so it would have to be added as an option first.
> 
> ```
> $ qemu-system-x86_64 -soundhw help
> Valid sound card names (comma separated):
> sb16        Creative Sound Blaster 16
> es1370      ENSONIQ AudioPCI ES1370
> ac97        Intel 82801AA AC97 Audio
> adlib       Yamaha YM3812 (OPL2)
> gus         Gravis Ultrasound GF1
> cs4231a     CS4231A
> hda         Intel HD Audio
> pcspk       PC speaker
> 
> -soundhw all will enable all of the above
> ```
> 
> There is a newer `-audiodev` flag. Then setting up pulseaudio is a headache all by itself, once the device is there.
> 
> I believe you want the "coreaudio" output device, for Mac ?
> 
> `-device intel-hda -device hda-output,audiodev=coreaudio`
> 
> Then once you have audio working in the VM, you need to map all the required volumes (for PA) into the container.
> 
> https://github.com/mviereck/x11docker/wiki/Container-sound:-ALSA-or-Pulseaudio

hi thanks for the reply ,  but where is the stored config of the qemu created by lima ?

**Reply by afbjorklund** [reply]
It's not really stored (as far as I know?), but for testing you could just use `ps`

I was thinking it could eventually using something like `audio:` in the YAML...

Maybe open a GitHub issue for it ?

**Comment by jandubois** [comment]
For testing you could try:

```shell
QEMU_SYSTEM_X86_64="qemu-system-x86_64 -device intel-hda -device hda-output,audiodev=coreaudio" limactl start
```

Then check if the audio device has been created, and try to setup PA.

If this works, then we can discuss providing an option to configure this, or add it by default (if the cost to do so is low).
